File: xwzx-news/src/api/favorite.js
/**
 * 收藏相关 API 接口
 */

// 检查单条新闻收藏状态
export async function checkFavorite(newsId) {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch(`/api/favorite/check?newsId=${newsId}`, {
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '查询收藏状态失败')
  return data.data.isFavorite
}

// 切换收藏（收藏/取消收藏）
export async function toggleFavorite(newsId) {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch(`/api/favorite/toggle?newsId=${newsId}`, {
    method: 'POST',
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '操作失败')
  return data.data.isFavorite
}

// 清空全部收藏
export async function clearFavorites() {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch('/api/favorite/clear', {
    method: 'DELETE',
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '清空失败')
  return data.message
}

// 获取收藏列表
export async function fetchFavorites() {
  const token = localStorage.getItem('user-token')
  if (!token) throw new Error('未登录')
  const res = await fetch('/api/favorite/list', {
    headers: { 'Authorization': token }
  })
  const data = await res.json()
  if (data.code !== 200) throw new Error(data.message || '获取收藏列表失败')
  return data.data
}

File: FastAPIUserService/routers/favorite.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from FastAPIUserService.config.db_config import get_db
from FastAPIUserService.crud.favorite import clear_all_favorite_news, get_favorite_news, is_favorite_news, toggle_favorite
from FastAPIUserService.models.users import User
from FastAPIUserService.schemas.favorite import FavoriteCheckResponse
from FastAPIUserService.utils.success_response import success_response
from FastAPIUserService.utils.auth import get_current_user


router = APIRouter(prefix="/api/favorite", tags=["favorite"])


@router.get("/check")
async def check_favorite(
    news_id: int = Query(..., alias="newsId"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    is_favorited = await is_favorite_news(db, user.id, news_id)
    return success_response(message="检查收藏状态成功", data=FavoriteCheckResponse(isFavorite=is_favorited))


@router.post("/toggle")
async def toggle_favorite_endpoint(
    news_id: int = Query(..., alias="newsId"),
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    is_fav = await toggle_favorite(db, user.id, news_id)
    return success_response(
        message="收藏成功" if is_fav else "已取消收藏",
        data={"isFavorite": is_fav}
    )


@router.get("/list")
async def list_favorites(
    user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    news_list = await get_favorite_news(db, user.id)
    result = [
        {
            "id": n.id,
            "title": n.title,
            "image": n.image,
            "author": n.author,
            "publish_time": n.publish_time,
            "views": n.views,
            "categoryId": n.category_id
        }
        for n in news_list
    ]
    return success_response(message="获取收藏列表成功", data=result)


# 清空收藏列表
@router.delete("/clear")
async def clear_favorite(
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_user)
):
    count = await clear_all_favorite_news(db, user.id)
    return success_response(message=f"清空了{count}收藏列表")